remote_attestation/rust/src/handshaker.rs

```diff
@@ -21,203 +21,153 @@
-    ///
-    /// Transitions [`ClientAttestationEngine`] state from [`Initializing`] to [`Attesting`] state.
-    pub fn create_client_hello(
-        mut self,
-    ) -> anyhow::Result<(Vec<u8>, ClientAttestationEngine<Attesting>)> {
-        let client_hello = ClientHello::new(&self.state.random);
-        let serialized_client_hello = client_hello
-            .serialize()
-            .context("Couldn't serialize client hello message")?;
-
-        // Update current transcript.
-        self.transcript
-            .append(&client_hello)
-            .context("Couldn't append client hello to the transcript")?;
-
-        let key_negotiator = KeyNegotiator::create(KeyNegotiatorType::Client)
-            .context("Couldn't create key negotiator")?;
-
-        let next_state =
-            Attesting::create(key_negotiator).context("Couldn't create attesting state")?;
-        let attestation_engine = ClientAttestationEngine {
-            behavior: self.behavior,
-            state: next_state,
-            transcript: self.transcript,
-        };
-        Ok((serialized_client_hello, attestation_engine))
+    fn is_completed(&self) -> bool {
+        match self.state {
+            HandshakerState::Completed(_) => true,
+            _ => false,
+        }
     }
-}
 
-impl ServerAttestationEngine<Initializing> {
-    pub fn new(behavior: AttestationBehavior) -> Self {
-        Self {
-            behavior,
-            state: Initializing::new(),
-            transcript: Transcript::new(),
+    fn get_encryptor(self) -> anyhow::Result<AeadEncryptor> {
+        match self.state {
+            HandshakerState::Completed(encryptor) => Ok(encryptor),
+            _ => Err(anyhow!("Handshake is not complete")),
         }
     }
+}
 
-    /// Responds to `ClientHello` message by creating a `ServerIdentity` message.
-    ///
-    /// `ServerIdentity` message contains an ephemeral public key for negotiating session keys.
-    /// If self attestation is enabled this message also provides necessary information to perform
-    /// remote attestation.
+impl ClientHandshaker {
+    /// Initializes the Remote Attestation handshake by creating a [`ClientHandshaker`] and a
+    /// serialized [`ClientHello`] message.
     ///
-    /// Transitions [`ServerAttestationEngine`] state from [`Initializing`] to [`Attesting`] state.
-    pub fn process_client_hello(
-        mut self,
-        serialized_client_hello: &[u8],
-    ) -> anyhow::Result<(Vec<u8>, ServerAttestationEngine<Attesting>)> {
-        let client_hello = ClientHello::deserialize(serialized_client_hello)
-            .context("Couldn't deserialize client hello message")?;
-
-        // Create server identity message.
-        let key_negotiator = KeyNegotiator::create(KeyNegotiatorType::Server)
+    /// Transitions [`ClientHandshaker`] state to [`HandshakerState::ExpectingServerIdentity`]
+    /// state.
+    pub fn create(behavior: AttestationBehavior) -> anyhow::Result<(Self, Vec<u8>)> {
```

Comment:
I know this follows a preexisting pattern, but returning a tuple of the handshaker and the client hello message at the same time does not feel like a very natural API. I would have expected a separate functions for initialising the handshaker and getting the client hello message, or perhaps return the client hello message when calling next_step the first time.

Reply:
Done